### PR TITLE
[Intel XPU] Enable (biased) grouped topk for xpu

### PR DESCRIPTION
**Pull request by gaopengff**
### Motivation
Added sycl kernel pass for biased_grouped_topk and grouped_topk on xpu device, which has better than previous default triton implementation.
### Modifications
Applied sycl kernel "moe_fused_gate" on xpu device, also added ut test cases.

---
### CI States

Latest PR Test (Base): :no_entry_sign: [Run #29563083843](https://github.com/sgl-project/sglang/actions/runs/29563083843)
Latest PR Test (Extra): :x: [Run #29563083691](https://github.com/sgl-project/sglang/actions/runs/29563083691)

### CODE REVIEW

File: python/sglang/srt/layers/moe/topk.py [hunk 1]
```diff
@@ -1006,6 +1008,66 @@ def grouped_topk_cpu(
...
+            ),
+            apply_routed_scaling_factor_on_output=bool(
+                apply_routed_scaling_factor_on_output
+            ),
+        )
+
+    # use default implementation
+    return grouped_topk_gpu(
+        gating_output,
+        gating_output,
```
Comment (mingfeima) [inline python/sglang/srt/layers/moe/topk.py]: 2 gating output?

File: python/sglang/srt/layers/moe/topk.py
Comment (mingfeima) [inline python/sglang/srt/layers/moe/topk.py:183]: 204 205 are duplicates.

File: python/sglang/srt/layers/moe/topk.py [hunk 2]
```diff
@@ -201,6 +201,8 @@ def fused_topk_deepseek(
         raise ImportError("mate is required for the biased grouped topk.")
 
     from sglang.srt.hardware_backend.musa.kernels.topk import topk_sigmoid, topk_softmax
+if _is_xpu:
+    from sgl_kernel import topk_sigmoid, topk_softmax
```
Comment (mingfeima) [inline python/sglang/srt/layers/moe/topk.py]:
```suggestion
```

File: python/sglang/srt/layers/moe/topk.py [hunk 3]
```diff
@@ -1006,6 +1008,66 @@ def grouped_topk_cpu(
...
+    correction_bias = torch.zeros(
+        gating_output.shape[1], dtype=gating_output.dtype, device=gating_output.device
+    )
+
+    if (
+        _is_xpu
+        # moe_fused_gate kernel ensures that num_experts/num_expert_group does not exceed MAX_VPT=32 now.
+        and experts_per_group <= 32
+        and is_power_of_two(num_experts)
+    ):
```
Comment (mingfeima) [inline python/sglang/srt/layers/moe/topk.py]:
```suggestion
    # moe_fused_gate kernel ensures that num_experts/num_expert_group does not exceed MAX_VPT=32 now.
    if (experts_per_group <= 32 and is_power_of_two(num_experts)):
```

File: python/sglang/srt/layers/moe/topk.py [hunk 4]
```diff
@@ -1562,6 +1624,30 @@ def biased_grouped_topk_gpu(
                 correction_bias,
             )
             return topk_values * scaling, topk_indices
+        elif (
+            _is_xpu
+            # moe_fused_gate kernel ensures that num_experts/num_expert_group does not exceed MAX_VPT=32 now.
+            and experts_per_group <= 32
+            and is_power_of_two(num_experts)
```
Comment (mingfeima) [inline python/sglang/srt/layers/moe/topk.py:1631]: same here

File: python/sglang/srt/layers/moe/topk.py [hunk 5]
```diff
@@ -1006,6 +1008,66 @@ def grouped_topk_cpu(
...
+    apply_routed_scaling_factor_on_output: Optional[bool] = False,
+    scoring_func: str = "softmax",
+):
+    num_experts = gating_output.shape[1]
+    experts_per_group = (
+        num_experts // num_expert_group if num_expert_group else num_experts
+    )
+    correction_bias = torch.zeros(
+        gating_output.shape[1], dtype=gating_output.dtype, device=gating_output.device
+    )
```
Comment (mingfeima) [inline python/sglang/srt/layers/moe/topk.py]: this is doing zeros every iteration

* put it inside if clause
* try to directly get rid of it, without creating the phony bias at all.

File: python/sglang/srt/layers/moe/topk.py [hunk 6]
```diff
@@ -1006,6 +1008,66 @@ def grouped_topk_cpu(
...
+        _is_xpu
+        # moe_fused_gate kernel ensures that num_experts/num_expert_group does not exceed MAX_VPT=32 now.
+        and experts_per_group <= 32
+        and is_power_of_two(num_experts)
+    ):
+        from sgl_kernel import moe_fused_gate
+
+        return moe_fused_gate(
+            gating_output.to(torch.float32),
+            correction_bias.to(torch.float32),
```
Comment (mingfeima) [inline python/sglang/srt/layers/moe/topk.py]: to will likely trigger memcpy, make it fused in the kernel. don't do an outplace copy.

**Review by mingfeima — CHANGES_REQUESTED** [review]

File: python/sglang/srt/layers/moe/topk.py (hunk 1 above)
Comment (gaopengff) [reply python/sglang/srt/layers/moe/topk.py]: Fixed, the first parameter should be "hidden_states"

File: python/sglang/srt/layers/moe/topk.py
Comment (gaopengff) [reply python/sglang/srt/layers/moe/topk.py:183]: Removed them.

File: python/sglang/srt/layers/moe/topk.py (hunk 3 above)
Comment (gaopengff) [reply python/sglang/srt/layers/moe/topk.py]: Done.

File: python/sglang/srt/layers/moe/topk.py (hunk 5 above)
Comment (gaopengff) [reply python/sglang/srt/layers/moe/topk.py]: I have a PR to make bias optional for kernel "moe_fused_gate", which will solve this problem. https://github.com/sgl-project/sgl-kernel-xpu/pull/301.
I also remove this phony bias allocation here.

File: python/sglang/srt/layers/moe/topk.py (hunk 6 above)
Comment (gaopengff) [reply python/sglang/srt/layers/moe/topk.py]: Actually this datatype setting is from cuda. Cuda uses fp32 for gating_output type here [cuda topk](https://github.com/sgl-project/sglang/blob/v0.5.14/python/sglang/srt/layers/moe/topk.py#L1281). Also, the moe_fused_kernel will assert bias.type equal to input type [moe_fused_gate.cu](https://github.com/sgl-project/sglang/blob/v0.5.14/sgl-kernel/csrc/moe/moe_fused_gate.cu#L388)
If we want to make it aligned to cuda, I think this bf16->fp32 copy is not avoidable.

File: python/sglang/srt/layers/moe/topk.py (hunk 4 above)
Comment (gaopengff) [reply python/sglang/srt/layers/moe/topk.py:1631]: Unlike previous, I think this "_is_xpu" is necessary here. This dispatch is in a general method "biased_grouped_topk", the former one is in "grouped_topk_xpu".

File: python/sglang/srt/layers/moe/topk.py (hunk 6 above)
Comment (mingfeima) [reply python/sglang/srt/layers/moe/topk.py]: we can fix this later, we can update kernel to make it accepts both float32 gating_output and bfloat16 gating_outputs and skip the outplace `to`. the dtype conversion can be fused in kernel.

File: python/sglang/srt/layers/moe/topk.py
Comment (mingfeima) [inline python/sglang/srt/layers/moe/topk.py]: what is the value of `scaling` when `apply_routed_scaling_factor_on_output` is false?

Comment (gaopengff) [reply python/sglang/srt/layers/moe/topk.py]: It should be "routed_scaling_factor" and it's a bug here. Since now topk_sigmoid would accept scaling factor, we can pass in it. I just fixed it.

File: python/sglang/srt/layers/moe/topk.py (hunk 6 above)
Comment (gaopengff) [reply python/sglang/srt/layers/moe/topk.py]: This kernel supports more dtype(bfloat16, float16) for gating_output at [sycl moe_fused_gate](https://github.com/sgl-project/sgl-kernel-xpu/blob/main/src/sycl/MoE_fused_gate.cpp#L28). We can keep it's original dtype for xpu pass for now. Also we could force kernel always calculate scores through fp32 when type T is other dtypes.  What's your suggestion?

**Review by mingfeima — APPROVED** [review]

File: python/sglang/srt/layers/moe/topk.py (hunk 6 above)
Comment (mingfeima) [reply python/sglang/srt/layers/moe/topk.py]: topk is a tricky operator because this will effect which experts to use. so tiny little error might introduce big diff here. we can stick to the current impl as long as accuracy tests passes at the current stage.
